README.md
# gravField
[Node.js](http://nodejs.org/) Backend for Multiplayer Functionality - HTML5 Canvas Interaction - [Socket.io](https://socket.io/) to Hold it all together

<img src="https://raw.githubusercontent.com/kevindweb/nodeGame/master/views/fontsAndImages/star.png" width=100 alt="gravFieldLogo"/>

## Downloading Global Variables

[Download NodeJS](https://nodejs.org/en/download/)
```
# npm and node commands installed
npm install -g nodemon
```

## Running Local 
```
cd 'pathToFolder'
npm install #installs all necessary packages
nodemon
```
Your local version of gravField should now be running on [localhost:5000](http://localhost:5000/).



To be completed by late June 2017




changes.txt
Necessary changes
  - we need to add a server case. When the server goes down, all games and sockets
    end and the pages reload. This will allow all data to be refreshed, including
    user placement, usernames, and overall user activity
  - rules modal - there is an logic issue with the panels - what needs to happen is
    when one panel is open, and the player clicks elsewhere, the panel should close.
    This happens sometimes, but not does not work every time. 


views/clientScripts/chat.js
$(document).ready(function(){
  // chat functions
  $(document).on('click', '.panel-heading span.icon_minim', function (e) {
      var $this = $(this);
      if (!$this.hasClass('panel-collapsed')) {
          $this.parents('.panel').find('.panel-body').slideUp();
          $this.addClass('panel-collapsed');
          $this.removeClass('glyphicon-minus').addClass('glyphicon-plus');
      } else {
          $this.parents('.panel').find('.panel-body').slideDown();
          $this.removeClass('panel-collapsed');
          $this.removeClass('glyphicon-plus').addClass('glyphicon-minus');
      }
  });
  $(document).on('focus', '.panel-footer input.chat_input', function (e) {
      var $this = $(this);
      if ($('#minim_chat_window').hasClass('panel-collapsed')) {
          $this.parents('.panel').find('.panel-body').slideDown();
          $('#minim_chat_window').removeClass('panel-collapsed');
          $('#minim_chat_window').removeClass('glyphicon-plus').addClass('glyphicon-minus');
      }
  });
  $(document).on('click','.icon_close',function(){
    $("#chatHolder").hide('slow', function(){ $("#chatHolder").remove();});
  });
  $("#btn-input").keyup(function(e){
    if(e.keyCode == 13){
      $("#btn-chat").click();
    }
  });
  $("#btn-chat").click(function(){
    var value = $("#btn-input").val();
    if (!value.replace(/\s/g, '').length){
      $("#btn-input").val("");
      $("#btn-input").attr("placeholder","Empty msg");
    } else{
      sendMessage('Kevin Deems',value);
      $("#btn-input").val("");
      $("#btn-input").attr("placeholder","Sent");
      setTimeout(function(){
        $("#btn-input").attr("placeholder","Enter msg");
      },3000);
    }
  });
});


views/clientScripts/game.js
$(document).ready(function(){
  var socket = io();
  var ourUsername;
  var ourUserId;
  var truthy = false; // this is a boolean that switches when we log in or die
  var x = document.body;
  var ourTimer,setTimer;
  // socket listeners
  // show current player count
  socket.on('playerCount',function(data){
    $("#playerCount").html(data.count);
  });
  // username was logged
  socket.on('usernameCreated',function(data){
    if(!data.error){
      document.getElementById("btn-input").disabled=false;
      document.getElementById("btn-input").placeholder="Enter msg";
      $("#smallTitle").html("GravField: "+data.name);
      ourUserId = data.id;
      window.ourUserId = ourUserId;
      document.getElementsByClassName("container")[0].style = 'opacity:0;';
      $(".glyphicon-remove").attr("style","");
      $("#playerInfo").attr("style","");
      setTimeout(function(){
        document.getElementsByClassName("container")[0].style = 'display:none;';
      },300);
      truthy = true;
      setTimeout(function(){
        window.createCanvasElement();
      },301);
      return;
    } else{
      $("#userName").attr("placeholder",data.error);
    }
  });
  socket.on('playerData',function(data){
    window.xPos = data.x;
    window.yPos = data.y;
    // leaderboards
    window.rankList = data.rankList;
  });
  socket.on('newMessage',function(data){
    if(data.userName==ourUsername){
      receiveMyMessage(data.message,data.userName);
    } else{
      receiveOtherMessage(data.message,data.userName);
    }
  });
  socket.on('emailSent',function(data){
    if(!data.error){
      $("#formName").val("");
      $("#formEmail").val("");
      $("#formMessage").val("");
      $("#formMessage").attr("placeholder",data.res);
      $("#human").val("");
    } else{
      $("#human").val(data.error);
    }
  });
  socket.on('blackout',function(data){
    if(truthy){ // if our game has started
      window.onPitchBlack(data.length);
    }
  });
  socket.on('countDownCount',function(data){
    $("#countdownSpan").html(data.time);
    clientCountDown(data.time);
  });
  // end of socket listeners
  function colorTest(){
  	var num1 = Math.floor(Math.random()*256);
  	var num2 = Math.floor(Math.random()*256);
  	var num3 = Math.floor(Math.random()*256);
  	x.style = "background-color: rgb("+num1+","+num2+","+num3+")";
    if(!truthy){
      setTimeout(colorTest,2000);
    } else{
      x.style = "background-color: rgb(255,255,255)";
    }
  }
  colorTest();
  // showing countdown clock
  function clientCountDown(timeLeft){
    ourTimer = timeLeft;
    setTimer = setInterval(clientCountDownStart,1000);
  }
  function clientCountDownStart(){
    if(ourTimer=='1:00'){
      ourTimer = 59;
    } else if(ourTimer<=0){
      clearInterval(setTimer);
      setTimeout(function(){
        ourTimer = '1:00';
        setTimer = setInterval(clientCountDownStart,1000);
      },3000);
    } else{
      ourTimer--;
    }
    $("#countdownSpan").html(ourTimer);
  }
  // adding points
  window.updatePoints = function(currPoints){
    $("#scoreSpan").html(currPoints);
  }
  // assiging username
  $("#userName").keyup(function(e){
    if(e.keyCode == 13){
      $("#loginButton").click();
    }
  });
  $("#loginButton").click(function(){
    var value = $("#userName").val();
    $("#userName").val("");
    if(!value.replace(/\s/g, '').length){
      $("#userName").attr("placeholder","real usrname pls");
      return;
    }
    if(value.split("").length>12){
      $("#userName").attr("placeholder","usrname < 10 pls");
    } else{
      ourUsername = value;
      window.ourUsername = ourUsername;
      usernameCreate(value);
    }
  });
  function usernameCreate(name){
    socket.emit('usernameCreate',{username:name});
  }
  // getting data about the player
  window.playerData = function(){
    socket.emit('playerData');
  }
  // sending message to server
  window.sendMessage = function(usrname,msg){
    socket.emit('sendMessage',{userName:usrname,message:msg});
  }
  /* there are two receive functions
  one for other users messages and
  the other for your message to be logged */
  function receiveMyMessage(msg,user){
    var box1 = document.getElementById("messageBase");
    var box2 = document.createElement('div');
    box2.className = "row msg_container base_sent";
    var box3 = document.createElement('div');
    box3.className = "col-md-10 col-xs-10";
    var box4 = document.createElement('div');
    box4.className = "messages msg_sent";
    var newmsg = document.createElement('p');
    newmsg.innerHTML = msg;
    var usrBox = document.createElement('time');
    usrBox.innerHTML = user;
    box4.appendChild(newmsg);
    box4.appendChild(usrBox);
    box3.appendChild(box4);
    box2.appendChild(box3);
    box1.appendChild(box2);
  }
  function receiveOtherMessage(msg,user){
    var box1 = document.getElementById("messageBase");
    var box2 = document.createElement('div');
    box2.className = "row msg_container base_receive";
    var box3 = document.createElement('div');
    box3.className = "col-md-10 col-xs-10";
    var box4 = document.createElement('div');
    box4.className = "messages msg_receive";
    var newmsg = document.createElement('p');
    newmsg.innerHTML = msg;
    var usrBox = document.createElement('time');
    usrBox.innerHTML = user;
    box4.appendChild(newmsg);
    box4.appendChild(usrBox);
    box3.appendChild(box4);
    box2.appendChild(box3);
    box1.appendChild(box2);
  }
  // modal for the rules button
  // dropdown animation onclick
  $('.dropdown').on('show.bs.dropdown', function (e) {
      $(this).find('.dropdown-menu').first().stop(true, true).slideDown();
  });
  // dropdown slide up after click
  $('.dropdown').on('hide.bs.dropdown', function (e) {
      $(this).find('.dropdown-menu').first().stop(true, true).slideUp();
  });
  // modal for the message button
  $("#human").keydown(function(){
       if(document.getElementById("human").value!=55){
         var button = document.getElementById("submitMessage");
         button.disabled = true;
       } else{
         var button = document.getElementById("submitMessage");
         button.disabled = false;
       }
     });
  $("#human").keyup(function(){
     if(document.getElementById("human").value!=55){
       var button = document.getElementById("submitMessage");
       button.disabled = true;
     } else{
       var button = document.getElementById("submitMessage");
       button.disabled = false;
     }
  });
  $("#submitMessage").attr("disabled","disabled");
  $("#submitMessage").click(function(e){
     e.preventDefault();
     message = {};
     message.from = '"'+$("#formName").val()+'" <gravfieldgame@gmail.com>';
     message.to = 'gravfieldgame@gmail.com';
     message.subject = 'Regarding gravField';
     message.html = '<h3>My name is:</h3><p>'+$("#formName").val()+'</p><br><h3>My message is:</h3><p>'+$("#formMessage").val()+'</p><br><h3>My email is:</h3><p>'+$("#formEmail").val()+'</p>';
  	 socket.emit('sendEmail',message);
  });
  // player dies, data is reset
  window.playerDeath = function(){
    $("#canvasElement").hide('slow', function(){ $("#canvasElement").remove();});
    var playAgain = document.createElement('button');
    playAgain.id = "playAgainButton";
    playAgain.style = "position: absolute; left: 100px; top:100px;";
    $(playAgain)
      .addClass("btn btn-primary btn-md")
      .html("Play Again?");
    $(document.body).append(playAgain);
    socket.emit('playerDeath');
    $("playAgainButton").click(function(){
      socket.emit('playAgain');
      window.createCanvasElement();
    });
  }
});


views/clientScripts/canvas.js
window.createCanvasElement = function(){
  // there will be a max size of player, but not max score
  var c = document.createElement('canvas'),ctx,truthyColor = false,truthy=true,fList=[],myReq,myReqCount=0,points=0,spaceTruth=false,keyMap={},colorInterval,sizeN;
  // remember to set truthy equal to true when player loses
  c.id = 'canvasElement';
  c.width = $(window).width();
  c.height = $(window).height()-20;
  colorInterval = setInterval(colorTest,2000);
  function colorTest(){
    if(truthyColor){
      clearInterval(colorInterval);
      c.style = "background-color: rgb(0,0,0)";
      setTimeout(function(){
        colorInterval = setInterval(colorTest,2000);
      },3000);
      return;
    }
  	var num1 = Math.floor(Math.random()*256);
  	c.style = "background-color: rgb("+num1+","+num1+","+num1+")";
  }
  colorTest();
  ctx = c.getContext('2d');
  // canvas movement
  var player = {
    x:40,
    y:40,
    spdX:20,
    spdY:20,
    width:30,
    height:30,
    name:'P',
  }
  ctx.font = '40px Arial';
  $(document).click(function(){
    if(truthy){
      truthy = false;
      cancelAnimationFrame(myReq);
    } else{
      truthy = true;
      myReq = requestAnimationFrame(updateNow);
    }
  });
  function floatingCall(){
    var random = Math.floor(Math.random()*c.width);
    var thisF = {
      x:random,
      y:c.height,
      spdX:0,
      spdY:-5,
      width:10,
      height:10,
      name:'e',
      increment:1
    };
    fList.push(thisF);
  }
  function floating(){
    myReqCount++;
    if(myReqCount%20===0){
      floatingCall();
    }
    var l = fList.length;
    // first for loop deletes dot if not on screen
    // second moves the remaining dots in their prospective position
    var i=0;
    while(i<fList.length){
      if(fList[i].y<30||testCollision(fList[i],player)){
        fList.splice(i,1);
      } else{
        fList[i].y+=fList[i].spdY;
        ctx.save();
        ctx.fillStyle = 'blue';
        ctx.fillRect(fList[i].x-(fList[i].height/2),fList[i].y-(fList[i].width/2),fList[i].height,fList[i].width);
        ctx.restore();
        i++;
      }
    }
  }
  function update(q){
    if(q.y>=c.height-q.height/3){
        truthy = false;
        cancelAnimationFrame(myReq);
        playerDeath();
        return;
    }
    if(!spaceTruth){
        q.y+=q.spdY/5;
    }
    if(!truthyColor){
      ctx.drawImage(document.getElementById("clientImage"),q.x-(q.height/2),q.y-(q.width/2),q.height,q.width);
    }
  }
  $(document).keydown(function(e){
    var keyCode = e.keyCode;
    keyMap[keyCode] = true;
    switch (keyCode){
      case 32:
        spaceTruth = true;
        break;
      case 68: //d
        if(keyMap[87]){
          diagonal('++');
          break;
        } else if(keyMap[83]){
          diagonal('+-');
          break;
        }
        player.x+=player.spdX;
        break;
      case 83: //s
        if(keyMap[65]){
          diagonal('--');
          break;
        } else if(keyMap[68]){
          diagonal('+-');
          break;
        }
        player.y+=player.spdY;
        break;
      case 65: //a
        if(keyMap[83]){
          diagonal('--');
          break;
        } else if(keyMap[87]){
          diagonal('-+');
          break;
        }
        player.x-=player.spdX;
        break;
      case 87: //w
        if(keyMap[68]){
          diagonal('++');
          break;
        } else if(keyMap[65]){
          diagonal('-+');
          break;
        }
        player.y-=player.spdY;
        break;
    }
  });
  $(document).keyup(function(e){
    var keyCode = e.keyCode;
    switch (keyCode){
      case 32:
        spaceTruth = false;
        break;
    }
    keyMap[keyCode] = false;
  });
  function diagonal(sign){
    if(sign==='++'){
      player.x+=player.spdX;
      player.y-=player.spdY;
    } else if(sign==='--'){
      player.x-=player.spdX;
      player.y+=player.spdY;
    } else if(sign==='-+'){
      player.x-=player.spdX;
      player.y-=player.spdY;
    } else{
      player.x+=player.spdX;
      player.y+=player.spdY;
    }
  }
  // test for collision between players/enemies
  function testCollision(p1,p2){
    var rect1 = {
      x:p1.x-(p1.width/2),
      y:p1.y-(p1.height/2),
      height:p1.height,
      width:p1.width
    }, rect2 = {
      x:p2.x-(p2.width/2),
      y:p2.y-(p2.height/2),
      height:p2.height,
      width:p2.width
    };
    if(rect1.x<=rect2.x+rect2.width&&rect2.x<=rect1.x+rect1.width&&rect1.y<=rect2.y+rect2.height&&rect2.y<=rect1.y+rect1.height){
      points+=p1.increment;
      if(points<=300){
        animate(p1.increment/2);
      }
      window.updatePoints(points);
      return true;
    } else{
      return false;
    }
  }
  myReq = requestAnimationFrame(updateNow);
  function updateNow(){
    ctx.clearRect(0,0,c.width,c.height);
    if(truthyColor){
      ctx.save();
      ctx.fillStyle = "white";
      ctx.font = "30px arcFont";
      ctx.textAlign = "center";
      ctx.fillText("Black Hole",c.width/2,c.height/2);
      ctx.restore();
    }
    update(player);
    floating();
    if(truthy){
        myReq = requestAnimationFrame(updateNow);
    }
  }
  $(document.body).append(c).hide().show(800);
  // when player dies
  function playerLost(){
    window.playerDeath(window.ourUserId);
    cancelAnimationFrame(myReq);
  }
  function animate(size){
    if(size==1){
      player.height+=1;
      player.width+=1;
      return;
    }
    var animation = setInterval(function(){
      if(size<=0){
        clearInterval(animation);
        return;
      }
      player.width+=1;
      player.height+=1;
      size-=1;
    },250/size);
  }
  // every 1 minute, the screen goes black for
  // a small time, to add challenge to the game
  window.onPitchBlack = function(length){
    truthyColor = true;
    setTimeout(function(){
      truthyColor = false;
      colorTest();
    },length);
  }
  window.onresize = function(){
    $("#canvasElement").attr("height",$(window).height()-20);
    $("#canvasElement").attr("width",$(window).width());
  }
}


www/app.js
var express = require('express'),
    app = express(),
    path = require('path'),
    http = require('http'),
    nodemailer = require('nodemailer'),
    transporter = nodemailer.createTransport({service: 'gmail',auth: {user: 'gravfieldgame@gmail.com',pass: '!easy2Remember'}}),
    server = http.Server(app),
    viewPath = path.join(__dirname+'/../views'),
    jsPath = path.join(__dirname+'/../views/clientScripts'),
    cssPath = path.join(__dirname+'/../views/styles'),
    fontPath = path.join(__dirname+'/../views/fontsAndImages'),
    server = app.listen(8080),
    io = require('socket.io')(server,{}),
    randomList=[],
    playerList = [],
    truthy = false,
    connections=[],
    countDown,
    intervals,
    pitchBlackLength = 3000;

// initialize paths to scripts, fonts, and css
app.set('views',viewPath);
app.set('view engine','pug');
app.use('/scripts',express.static(jsPath));
app.use('/styles',express.static(cssPath));
app.use('/fonts',express.static(fontPath));

app.get('/',function(req,res){
  res.sendFile(path.join(__dirname+'/../views/render/index.html'));
});

//countDown until blackout
function setTimerInterval(){
  intervals = setInterval(countDownTimer,1000);
}
function countDownTimer(){
  if(!countDown){
    countDown = '1:00';
  } else{
    if(countDown=='1:00'){
      countDown = 59;
    } else if(countDown==1){
      clearInterval(intervals);
      countDown = null;
      setTimeout(function(){
        setTimerInterval();
      },pitchBlackLength);
      io.sockets.emit('blackout',{length:pitchBlackLength});
    } else{
      countDown--;
    }
  }
}
setTimerInterval();
// socket commands
io.sockets.on('connection', function (socket) {
    connections.push(socket);
    var player = function(id){
      var self = {
        x:250,
        y:250,
        id:id,
      };
      return self;
    }
    // default username is John Doe
    socket.username = 'John Doe';
    socket.emit('playerCount',{count:connections.length});
    function findRank(){
      // sort playerList array by total score
      // return array of top ten players
    }
    // sends a pitch black function to the sockets
    // every 60 seconds, the screen will go black for 4 seconds
    // function emitPitchBlack(){
    //   if(connections.length>1){
    //     setTimeout(function(){
    //       // after 60 seconds, send pitch black
    //       io.sockets.emit('pitchBlack');
    //       setTimeout(function(){
    //         // wait until pitch black is over, then start
    //         // timer again
    //         emitPitchBlack();
    //       },4000);
    //     },60*1000);
    //   } else{
    //     setTimeout(function(){
    //       // if there are not enough players online to
    //       // make it worth while, wait another minute and try again
    //       emitPitchBlack();
    //     },60*1000);
    //   }
    // }
    // emitPitchBlack();
    // all socket listeners
    socket.on('sendMessage',function(data){
      io.sockets.emit('newMessage',{message:data.message,userName:socket.username})
    });
    socket.on('usernameCreate',function(data){
      socket.username = data.username;
      var id;
      function getRandom(){
      	var myRandom = Math.random();
        // if we already have this random id - choose another
      	if(randomList.indexOf(myRandom==-1)){
      		randomList.push(myRandom);
      		id = myRandom;
      	} else{
      	   getRandom();
      	}
      };
      getRandom();
      socket.id = id;
      playerList[socket.id] = player(socket.id);
      socket.emit('usernameCreated',{error:null,name:data.username,id:socket.id});
      socket.emit('countDownCount',{time:countDown});
    });
    // when a user wants to send email
    socket.on('sendEmail',function(data){
      // send mail with defined transport object
      var mailOptions = data;
      transporter.sendMail(mailOptions, (error, info) => {
          if (error) {
              console.log(error.message);
              socket.emit('emailSent',{error:JSON.stringify(error)});
          } else{
            socket.emit('emailSent',{res:'Message was sent!'});
          }
      });
    });
    // request for player position and size, and game ranks
    socket.on('playerData',function(data){
      if(rankList==null){
        rankList = findRank();
      } else if(data.score>rankList[9]){
        rankList = findRank();
      }
      socket.emit('playerData',{x:socket.xPos,y:socket.yPos,rankList:rankList})
    });
    // when a player dies
    socket.on('playerDeath',function(){
      playerList.splice(playerList.indexOf(socket.id),1);
    });
    socket.on('playAgain',function(data){
      playerList[socket.id] = player(socket.id);
    });
    socket.on('disconnect', function (){
        connections.splice(connections.indexOf(socket),1);
        if(playerList[socket.id]){
          delete playerList[socket.id];
        }
    });
});
